proposals/csharp-10.0/parameterless-struct-constructors.md

```diff
@@ -0,0 +1,161 @@
+```csharp
+_ = default(NoConstructor);      // ok
+_ = default(PublicConstructor);  // ok: constructor ignored
+_ = default(PrivateConstructor); // ok: constructor ignored
+```
+
+Object creation expressions require the parameterless constructor to be accessible if defined.
+The parameterless constructor is invoked explicitly.
+_This is a breaking change if the struct type with parameterless constructor is from an existing assembly._
+```csharp
+_ = new NoConstructor();       // ok: initobj NoConstructor
+_ = new InternalConstructor(); // ok: call InternalConstructor::.ctor()
+_ = new PrivateConstructor();  // error: 'PrivateConstructor..ctor()' is inaccessible
+```
+
+A local or field of a struct type that is not explicitly initialized is zeroed.
+The compiler reports a definite assignment error for an uninitialized struct that is not empty. 
+```csharp
+NoConstructor s1;
+PublicConstructor s2;
+s1.ToString(); // error: use of unassigned local (unless type is empty)
+s2.ToString(); // error: use of unassigned local (unless type is empty)
+```
+
+Array allocation ignores any parameterless constructor and generates zeroed elements.
+_Should the compiler warn that the parameterless constructor is ignored? How would such a warning be avoided?_
+```csharp
+_ = new NoConstructor[1];      // ok
+_ = new PublicConstructor[1];  // ok: constructor ignored
+_ = new PrivateConstructor[1]; // ok: constructor ignored
+```
+
+Parameterless constructors cannot be used as parameter default values.
+_This is a breaking change if the struct type with parameterless constructor is from an existing assembly._
+```csharp
+static void F1(NoConstructor s1 = new()) { }     // ok
+static void F2(PublicConstructor s1 = new()) { } // error: default value must be constant
+```
+
+### Constraints
+The `new()` type parameter constraint requires the parameterless constructor to be `public` if defined (see [satisfying constraints](https://github.com/dotnet/csharplang/blob/master/spec/types.md#satisfying-constraints)).
+```csharp
+static T CreateNew<T>() where T : new() => new T();
+
+_ = CreateNew<NoConstructor>();       // ok
+_ = CreateNew<PublicConstructor>();   // ok
+_ = CreateNew<InternalConstructor>(); // error: 'InternalConstructor..ctor()' is not public
+_ = CreateNew<PrivateConstructor>();  // error: 'PrivateConstructor..ctor()' is not public
+```
+
+`new T()` is emitted as a call to `System.Activator.CreateInstance<T>()`, and the compiler assumes the implementation of `CreateInstance<T>()` invokes the `public` parameterless constructor if defined.
+
+There is a gap in type parameter constraint checking because the `new()` constraint is satisfied by a type parameter with a `struct` constraint (see [satisfying constraints](https://github.com/dotnet/csharplang/blob/master/spec/types.md#satisfying-constraints)).
+
+As a result, the following will be allowed by the compiler but `Activator.CreateInstance<InternalConstructor>()` will fail at runtime.
```

Comment:
Sorry for being late to the party. Can we avoid this problem by requiring `public` accessibility on the parameterless constructor?

Reply:
Unfortunately not really. The ecosystem today already contains `struct` with non-public ctors. That has to be accounted for in the design here. It would be possible for us to say we don't support them at a syntax level but we need to account for what happens when we see them in metadata. 

Given we have to account for them in metadata I would allow them in source otherwise it creates an odd friction point.

Reply:
Right, it makes sense that we still have to handle types in the wild. But are we helping users by making it easy to step into this gap?
If possible, let's bring this as an open question at LDM tomorrow.